src/vm/mod.rs

```diff
@@ -1,3 +1,80 @@
 // SPDX-License-Identifier: Apache-2.0
 
 mod linux;
+
+use crate::vm::linux::{
+    ioctl::Cmd,
+    types::{Capabilities, CpuidConfig},
+};
+use kvm_ioctls::{Error, Kvm, VmFd};
+
+// Defined in linux/arch/x86/include/uapi/asm/kvm.h
+const KVM_X86_TDX_VM: u64 = 2;
+
+/// Handle to the KVM VM file descriptor
+pub struct TdxVm(VmFd);
+
+impl TdxVm {
+    /// Create a new TDX VM with KVM
+    pub fn create(kvm_fd: Kvm) -> Result<Self, Error> {
+        let vm_fd = kvm_fd.create_vm_with_type(KVM_X86_TDX_VM)?;
+        Ok(TdxVm(vm_fd))
+    }
+
+    /// Retrieve information about the Intel TDX module
+    pub fn get_capabilities(&self) -> Result<TdxCapabilities, kvm_ioctls::Error> {
+        // configure the kernel facing capabilities structure
+        let caps = Capabilities::new();
+
+        // configure the kernel facing TDX command that will be wrapped by the
+        // MEMORY_ENCRYPT_OP ioctl
+        let mut cmd: Cmd = Cmd::from(&caps);
+
+        unsafe { self.0.encrypt_op(&mut cmd)? };
+
+        // MEMORY_ENCRYPT_OP ioctl was a success so we can return the TDX
+        // capabilities in a user-friendly way
+        Ok(TdxCapabilities {
+            fixed_attrs: FixedAttrs {
+                fixed0: caps.attrs_fixed0,
+                fixed1: caps.attrs_fixed1,
+            },
+            fixed_xfam: FixedXFAM {
+                fixed0: caps.xfam_fixed0,
+                fixed1: caps.xfam_fixed1,
+            },
+            supported_gpaw: caps.supported_gpaw,
+            cpuid_configs: Vec::from(caps.cpuid_configs),
+        })
+    }
+}
+
+/// Reflects the Intel TDX module capabilities and configuration and CPU
+/// capabilities
+#[derive(Debug)]
+pub struct FixedAttrs {
+    /// If any bit is a 0 in the bitmap, it must be 0 in a TD's ATTRIBUTES
+    pub fixed0: u64,
+    /// If any bit is a 1 in the bitmap, it must be 1 in a TD's ATTRIBUTES
+    pub fixed1: u64,
+}
+
+/// Determines the set of extended features available for use by the guest TD
+#[derive(Debug)]
+pub struct FixedXFAM {
+    /// If any bit is a 0 in the bitmap, it must be a 0 in the TD's XFAM
```

Comment:
What bitmap are you referring to here?

Reply:
You can find the documentation in [the ABI spec](https://www.intel.com/content/www/us/en/content-details/733579/intel-tdx-module-v1-5-abi-specification.html)  when  you search for `SYSINFO` which reports information about the system.  You can find more specific information about it in [the Base Architecture Spec](https://www.intel.com/content/www/us/en/content-details/733575/intel-tdx-module-v1-5-base-architecture-specification.html). Essentially, it is a 64bit mask that specifies additional features the TD can use. What features are associated with what bits can be found on page 94 of the Base Architecture Spec. I use the word "bitmap" here in the comment because in the ABI that is what the type is referred as.

Reply:
I will  come up with clearer documentation